kv/kvserver: TestReplicateQueueShouldQueueNonVoter failed

kv/kvserver.TestReplicateQueueShouldQueueNonVoter [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10303707?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10303707?buildTab=artifacts#/) on release-23.1 @ [14eebdf2af685fba00e12e8e5e92fc202c4a6c0b](https://github.com/cockroachdb/cockroach/commits/14eebdf2af685fba00e12e8e5e92fc202c4a6c0b):


```
          | github.com/stretchr/testify/assert.Eventually.func1
          | 	github.com/stretchr/testify/assert/external/com_github_stretchr_testify/assert/assertions.go:1737
          | runtime.goexit
          | 	GOROOT/src/runtime/asm_amd64.s:1594
        Wraps: (2) /Table/Max: lookup range unexpected error
        Wraps: (3)
          | (opaque error wrapper)
          | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
          | reportable 0:
          |
          | github.com/cockroachdb/cockroach/pkg/rpc/nodedialer.(*Dialer).dial
          | 	github.com/cockroachdb/cockroach/pkg/rpc/nodedialer/nodedialer.go:205
          | github.com/cockroachdb/cockroach/pkg/rpc/nodedialer.(*Dialer).DialInternalClient
          | 	github.com/cockroachdb/cockroach/pkg/rpc/nodedialer/nodedialer.go:167
          | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).NextInternalClient
          | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/transport.go:252
          | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).SendNext
          | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/transport.go:185
          | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendToReplicas
          | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:2163
          | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendPartialBatch
          | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1689
          | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).divideAndSendBatchToRanges
          | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1260
          | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).Send
          | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:881
          | github.com/cockroachdb/cockroach/pkg/kv.(*CrossRangeTxnWrapperSender).Send
          | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:223
          | github.com/cockroachdb/cockroach/pkg/kv.lookupRangeFwdScan
          | 	github.com/cockroachdb/cockroach/pkg/kv/range_lookup.go:331
          | github.com/cockroachdb/cockroach/pkg/kv.RangeLookup
          | 	github.com/cockroachdb/cockroach/pkg/kv/range_lookup.go:206
          | github.com/cockroachdb/cockroach/pkg/server.(*TestServer).LookupRange
          | 	github.com/cockroachdb/cockroach/pkg/server/testserver.go:1472
          | github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).LookupRange
          | 	github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:662
          | github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).LookupRangeOrFatal
          | 	github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:668
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestReplicateQueueShouldQueueNonVoter.func1
          | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replicate_queue_test.go:1621
          | github.com/stretchr/testify/assert.Eventually.func1
          | 	github.com/stretchr/testify/assert/external/com_github_stretchr_testify/assert/assertions.go:1737
          | runtime.goexit
          | 	GOROOT/src/runtime/asm_amd64.s:1594
        Wraps: (4) failed to connect to n3 at 127.0.0.1:45311
        Wraps: (5) grpc: refusing to dial; node is quiescing [code 7/PermissionDenied]
        Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *errbase.opaqueWrapper (4) *errutil.withPrefix (5) *status.Error
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestReplicateQueueShouldQueueNonVoter2875125237
--- FAIL: TestReplicateQueueShouldQueueNonVoter (67.87s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #103107 kv/kvserver: TestReplicateQueueShouldQueueNonVoter failed [C-bug C-test-failure O-robot T-kv todo-deprecate.branch-release-23.1.0]
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicateQueueShouldQueueNonVoter.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-28340

--- a/pkg/kv/kvserver/allocator/plan/util.go
+++ b/pkg/kv/kvserver/allocator/plan/util.go
@@ -36,6 +36,7 @@ func ReplicationChangesForRebalance(
 	addTarget, removeTarget roachpb.ReplicationTarget,
 	rebalanceTargetType allocatorimpl.TargetReplicaType,
 ) (chgs []kvpb.ReplicationChange, performingSwap bool, err error) {
+	rdesc, found := desc.GetReplicaDescriptor(addTarget.StoreID)
 	if rebalanceTargetType == allocatorimpl.VoterTarget && numExistingVoters == 1 {
 		// If there's only one replica, the removal target is the
 		// leaseholder and this is unsupported and will fail. However,
@@ -58,14 +59,29 @@ func ReplicationChangesForRebalance(
 		// when we know it's necessary, picking the smaller of two evils.
 		//
 		// See https://github.com/cockroachdb/cockroach/issues/40333.
-		chgs = []kvpb.ReplicationChange{
-			{ChangeType: roachpb.ADD_VOTER, Target: addTarget},
-		}
 		log.KvDistribution.Infof(ctx, "can't swap replica due to lease; falling back to add")
+
+		// Even when there is only 1 existing voter, there may be other replica
+		// types in the range. Check if the add target already has a replica, if so
+		// it must be a non-voter or the rebalance is invalid.
+		if found && rdesc.Type == roachpb.NON_VOTER {
+			// The receiving store already has a non-voting replica. Instead of just
+			// adding a voter to the receiving store, we *must* promote the non-voting
+			// replica to a voter.
+			chgs = kvpb.ReplicationChangesForPromotion(addTarget)
+		} else if !found {
+			chgs = []kvpb.ReplicationChange{
+				{ChangeType: roachpb.ADD_VOTER, Target: addTarget},
+			}
+		} else {
+			return nil, false, errors.AssertionFailedf(
+				"invalid rebalancing decision: trying to"+
+					" move voter to a store that already has a replica %s for the range", rdesc,
+			)
+		}
 		return chgs, false, err
 	}
 
-	rdesc, found := desc.GetReplicaDescriptor(addTarget.StoreID)
 	switch rebalanceTargetType {
 	case allocatorimpl.VoterTarget:
 		// Check if the target being added already has a non-voting replica.

